docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -1,184 +1,224 @@
+
+class Derived : public Base {
+  public:
+  KOKKOS_FUNCTION
+  void Bar() override {}
+};
+
+int main(int argc, char *argv[])
+{
+  Kokkos::ScopeGuard kokkos(argc, argv);
+
+  // create
+  auto deviceInstanceMemory = Kokkos::kokkos_malloc(sizeof(Derived)); // allocate memory on device
+  Kokkos::parallel_for("initialize", 1, KOKKOS_LAMBDA (const int i) {
+    new (static_cast<Derived*>(deviceInstanceMemory)) Derived(); // initialize on device
+  });
+  Kokkos::fence("wait for initialize");
+  Base* deviceInstance = static_cast<Derived*>(deviceInstanceMemory); // declare on this memory
+
+  // use
+  Kokkos::parallel_for("myKernel", 10, KOKKOS_LAMBDA (const int i) {
+      deviceInstance->Bar();
+  });
+  Kokkos::fence();
+
+  // cleanup
+  Kokkos::parallel_for("destroy", 1, KOKKOS_LAMBDA (const int i) {
+    deviceInstance->~Base(); // destroy on device
+  });
+  Kokkos::fence("wait for destroy");
+  Kokkos::kokkos_free(deviceInstanceMemory); // free
+}
 ```
 
+We first use the `KOKKOS_FUNCTION` macro to make the methods callable from a kernel.
+When creating the instance, note that we introduce a distinction between the *memory* that the it uses, and the actual instantiated *object*.
+The construct is done on the device, within a single-iteration `parallel_for`, using placement new.
+Since the kernel does not have a return type, we use a static cast to associate the object with the memory allocation.
+
+Like with other uses of `new`, we need to later free the memory.
+The destruct is done on the device, again with a single-iteration `parallel_for`.
+
 This code is extremely ugly, but leads to functional virtual function calls on the device. The Vpointer now points to the device Vtable.
+Remember that those virtual functions cannot be called on the host anymore!
 
 ![VPointerToDevice](./figures/VirtualFunctions-VPointerToDevice.png)
 
-Note that like with other uses of `new`, you need to later `free` the memory.
 For a full working example, see [the example in the repo](https://github.com/kokkos/kokkos/blob/master/example/virtual_functions/main.cpp).
 
-## Complications and Fixes
-
-The first problem people run into with this is that they want to initialize some fields or nested classes based on host data before moving data down to the device
-
-```c++
-Base* hostInstance = new Derived(); // allocate and initialize host
-hostInstance->setAField(someHostValue);
-Base* deviceInstance; // cudaMalloc'd to allocate
-cudaMemcpy(deviceInstance, hostInstance); // to initialize the deivce
-Kokkos::parallel_for(... {
-  // use deviceInstance
-});
-```
+## What if I need a setter that works with host values?
 
-We can't translate this easily, the naive translation would be
+The first problem people run into with this is that they want to initialize some fields based on host data, with a setter which is *not* a virtual function.
+Calling this setter on the device would crash if the host data is not easily copyable on the device (e.g. for a small array).
```

Comment:
```suggestion
Calling this setter on the device would crash if the host data is not easily copyable on the device.
```

I don't understand the difficulty with small arrays ?

Reply:
This is a remaining of the last iteration. You could do, as the text says:

```cpp
Kokkos::parallel_for("myLoop", 1, KOKKOS_LAMBDA (const int i) {
  deviceInstance.setAField(hostValue);
});

But if `hostValue` is not a scalar type (like the pointer of an array), the code will crash.

Reply:
I see, I don't feel that mentioning "small array" helps in the end

Reply:
there could be multiple things wrong here. 

- The lambda capture could result in an invalid instance on the device.
- The type itself could not be copyable or assignable on the device.
- Side effects could happen in at unexpected times, because async kernel run.

So maybe we could write:
```suggestion
Calling this setter on the device would crash if `hostValue` does not form a valid type on the device 
```

Reply:
I think it would be more pedagogical to explicit what could go wrong. When I read the page before this PR, I didn't get why the capture of the lambda wouldn't be enough.

Reply:
I tried another formulation based on @JBludau's remark.